[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1bv9Di1PmgvNHiL8Ue2LWTjekvfYXEQ8k?usp=sharing)

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.214
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4400.42
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

69 1337
1406

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Overwriting output1.txt


In [ ]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
9

Overwriting output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

1488 420
1908


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_full.cpp


##### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

10
1 2 3 4 5 6 7 8 9 10
90


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Overwriting input0.txt


In [ ]:
%%writefile output0.txt
0

Overwriting output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Overwriting input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {0..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
10
200000 23 200000 0 45 1 4 6 123 32131

Overwriting input4.txt


In [ ]:
200000 * 200000

40000000000

In [ ]:
%%writefile output4.txt
40000000000

Overwriting output4.txt


In [ ]:
%%writefile input5.txt
5
3 1 200000 45 20000

Overwriting input5.txt


In [ ]:
200000 * 20000

4000000000

In [ ]:
%%writefile output5.txt
4000000000

Overwriting output5.txt


In [ ]:
%%writefile input6.txt
15
199999 234 0 234 3000 124511 400000 4523 7845 121314 89677 453123 523213 7453 0

Overwriting input6.txt


In [ ]:
453123 * 523213

237079844199

In [ ]:
%%writefile output6.txt
237079844199

Overwriting output6.txt


In [ ]:
%%writefile input7.txt
2
900000 5

Overwriting input7.txt


In [ ]:
900000 * 5

4500000

In [ ]:
%%writefile output7.txt
4500000

Overwriting output7.txt


In [ ]:
%%writefile input8.txt
19
28543 11213 7663 35730 28443 14637 19039 25441 20156 16505 30649 23923 25435 18913 29108 25409 17174 143 11733

Overwriting input8.txt


In [ ]:
35730 * 30649

1095088770

In [ ]:
%%writefile output8.txt
1095088770

Overwriting output8.txt


In [ ]:
%%writefile input9.txt
50
59918 10147 58122 15353 69460 86114 18171 45377 79182 68269
33235 50458 37455 14672 78008 28725 41017 56580 49624 29429
23106 12679 29952 26702 84626 25185 45291 89627 38901 23801
53468 80883 68361 72442 87880 34449 17123 56210 52144 46789
62951 12690 46326 85744 11059 36021 49412 11731 21735 33212

Overwriting input9.txt


In [ ]:
89627 * 87880

7876420760

In [ ]:
%%writefile output9.txt
7876420760

Overwriting output9.txt


In [ ]:
%%writefile input10.txt
60
16695 50605 42530 57855 24236 24347 36570 58107 25949 26939
10769 54063 66323 23606 22007 37126 51412 14381 43186 12283
34525 55947 52947 66780 14485 41951 63416 32926 27342 64352
37077 27768 27124 54245 63911 28044 12350 64340 57079 51652
38353 17289 45777 45391 22221 60764 26161 36591 12275 49150
123888 129855 116245 96360 107581 124692 133711 82847 116602 103602

Overwriting input10.txt


In [ ]:
17363041905

17363041905

In [ ]:
%%writefile output10.txt
17363041905

Overwriting output10.txt


In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {0..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
#Running tests cases. Check "/files" for "outputN.txt" where N is the Nth output for the Nth "inputN.txt" (e.g. input4.txt/output4.txt)
!chmod 744 stress_test.sh
!./stress_test.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	5m0.187s
user	4m57.153s
sys	0m0.088s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m57.564s
user	4m55.367s
sys	0m0.059s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound

Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 ////INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);
    execints++;

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right) \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

571


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$

In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

601


For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt
n= 10

572


For the input11.txt
```
200000
1 2 ..  199999 200000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

In [ ]:
# Estimated time by instruction
295/200001000021


1.4749926248820013e-09

In [ ]:
# Estimated number of instructions by second
1 / (295/200001000021)

677969491.5966102

# Estimate the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.886s
user	0m0.869s
sys	0m0.005s


## Python

In [ ]:
%%writefile max_pairwise_product_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(n):
            if i != j:
                aux_product = l[i] * l[j]
                if aux_product > max_product:
                    max_product = aux_product
    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Overwriting max_pairwise_product_py.py


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("11000\n")
        for i in range(1, 11001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
# Verify #Not surprising Python version is slower...
%%time
!python max_pairwise_product_py.py < input14.txt

120989000
CPU times: user 187 ms, sys: 35.2 ms, total: 222 ms
Wall time: 21.5 s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound

Overwriting max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input11.txt

39999800000

real	2m36.583s
user	2m26.080s
sys	0m0.179s


In [ ]:
%%writefile max_pairwise_product_n2_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(i + 1, n):
            max_product = max(max_product, l[i] * l[j])

    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Overwriting max_pairwise_product_n2_py.py


The number of instructions executed by the algorithm with a list of size $n$ is  

$$T (n) = 11 + 7n + \left[2(n-1)n+1, 2.5(n-1)n\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```

$$T(10) \in [262, 306] $$

In [ ]:
n = 10
print(11 + 7*(10) + 2*(10)*(9) + 1, 11 + 7*(10) + 2.5*(10)*(9))

262 306.0


For the input11.txt
```
200000
1 2 . . . 199999 200000
```

$$T(200000) \in [80001000012, 100000900011] $$

In [ ]:
print(11 + 7*(200000) + 2*(200000)*(199999) + 1, 11 + 7*(200000) + 2.5*(200000)*(199999))

80001000012 100000900011.0


In [ ]:
#ins
estimated_time = 195 / 100000900009
estimated_time

1.9499824499824516e-09

In [ ]:
#ins per sec (aprox.)
ins_per_sec =  1 / estimated_time
ins_per_sec

512825128.2512821

###Correctness tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long.cpp -o max_pairwise_product_long

for i in {0..10}
do
    ./max_pairwise_product_long < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_n2_py.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

###Estimated maximum that can be solved in 1 second
$$(11 + 7n + 2.5(n-1)n)*1.47*10^{-9} = 1s$$
$$ n \approx 16400 $$

In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 16400 <<"\n";  // n
 for (int i=1; i <16401; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

268943600

real	0m0.904s
user	0m0.891s
sys	0m0.000s


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("16400\n")
        for i in range(1, 16401):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
#Slower. Not a surprise.
%%time
!python max_pairwise_product_n2_py.py < input14.txt

268943600
CPU times: user 195 ms, sys: 28.6 ms, total: 224 ms
Wall time: 20 s


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    int n;                                                   //1
    std::cin >> n;                                           //1
    std::vector<int> v(n);                                   //n
    for(int i = 0; i < n; i++){                              //1 + 2n
        std::cin >> v[i];                                    //n
    }

    std::sort(v.begin(), v.end());                          //nlogn
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';   //2
    return 0;
}

Writing max_pairwise_product_sort.cpp


In [ ]:
%%writefile max_pairwise_product_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())                                        #1
    l = list(map(int, input().split()))                     #5n
    l.sort()                                                #nlogn at least
    print(l[-1] * l[-2])                              #2

Overwriting max_pairwise_product_sort.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!./max_pairwise_product_sort < input1.txt   #Original array in input1.txt increasing order [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
!./max_pairwise_product_sort < input2.txt   #Original array in input2.txt decreasing order [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

90
90


In [ ]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input3.txt

40000000000

real	0m0.006s
user	0m0.001s
sys	0m0.002s


In [ ]:
%%time
#Given in secs. *10^-3 ms
!python max_pairwise_product_sort.py < input3.txt

40000000000
CPU times: user 5.02 ms, sys: 1.03 ms, total: 6.05 ms
Wall time: 105 ms


20 stress tests

In [ ]:
%%writefile stress_test_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 49 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::sort(v.begin(), v.end());
        std::cout << "n: " << n << " Product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }
    return 0;
}

Writing stress_test_cpp.cpp


In [ ]:
!g++ stress_test_cpp.cpp -o stress_test_cpp
!./stress_test_cpp

n: 17 Product: 850976
n: 25 Product: 857451
n: 43 Product: 980036
n: 10 Product: 692094
n: 30 Product: 839816
n: 48 Product: 956363
n: 9 Product: 475580
n: 46 Product: 954088
n: 44 Product: 902850
n: 17 Product: 981034
n: 40 Product: 928812
n: 7 Product: 267968
n: 42 Product: 991020
n: 27 Product: 908160
n: 23 Product: 990016
n: 2 Product: 141372
n: 4 Product: 418500
n: 11 Product: 557996
n: 42 Product: 963270
n: 17 Product: 690616


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_sort_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                                                   //1
    ins++;
    std::cin >> n;                                           //1
    ins++;
    std::vector<int> v(n);                                   //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){                              //1 + 2n
        ins += 2;
        std::cin >> v[i];                                    //n
        ins++;
    }

    std::sort(v.begin(), v.end());                           //nlogn at least (nlogn time complexity)
    ins += n * log2(n);
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';    //3
    ins += 3;
    std::cout << "Executed instructions: " << ins << '\n';   // 7 + 4n + ins(nlogn sort). Where ins(nlogn sort) are, at least, nlogn.
    return 0;
}

Writing max_pairwise_product_sort_ins.cpp


In [ ]:
!g++ max_pairwise_product_sort_ins.cpp -o max_pairwise_product_sort_ins
!./max_pairwise_product_sort_ins < input1.txt
!./max_pairwise_product_sort_ins < input2.txt
!./max_pairwise_product_sort_ins < input3.txt

90
Executed instructions: 79
90
Executed instructions: 79
40000000000
Executed instructions: 45


In [ ]:
!time ./max_pairwise_product_sort_ins < input11.txt

39999800000
Executed instructions: 4321934

real	0m0.130s
user	0m0.124s
sys	0m0.002s


In [ ]:
%%writefile max_pairwise_product_sort_ins.py
#Python3
import math

if __name__ == "__main__":
    ins = 0
    n = int(input())                       #1
    ins += 1
    l = list(map(int, input().split()))    #5n (n input read, n mappings to int, n iterable to list, n comparisons, n splits through input)
    ins += 5 * n
    l.sort()                               #nlogn at least
    ins += n * math.log2(n)
    print(l[-1] * l[-2])             #2
    ins += 2
    print(f"Executed instructions: {int(ins)}\n")                             #3 + 5n + ins(nlogn timsort) where ins(nlogn sort) are, at least, nlogn.

Overwriting max_pairwise_product_sort_ins.py


In [ ]:
!python max_pairwise_product_sort_ins.py < input1.txt
!python max_pairwise_product_sort_ins.py < input2.txt
!python max_pairwise_product_sort_ins.py < input3.txt

90
Executed instructions: 86

90
Executed instructions: 86

40000000000
Executed instructions: 48



In [ ]:
%%time
!python max_pairwise_product_sort_ins.py < input11.txt

39999800000
Executed instructions: 4521931

CPU times: user 9.06 ms, sys: 1.01 ms, total: 10.1 ms
Wall time: 212 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
#following result of input3.txt. #lower bound. Needs to know the exact number of instructions in each sorting algorithm. Just counts the ones in the nlogn step.
#Estimaded time by instruction and number of instructions
estimated_time = 0.009 / 4521931
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

1.9903001615902584e-09
502436777.77777785


###Estimated maximum that can be solved in 1 second
$$(5 + 4n + 5nlog(n))*1.47*10^{-9} = 1s \ \ \ given \ that \ nlogn \ at \ least$$
$$ n \approx 8'000.000 $$

In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 8000000 <<"\n";  // n
 for (int i=1; i <8000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest

In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input13.txt

3599999940000000

real	0m52.249s
user	0m50.922s
sys	0m0.413s


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input16.txt", "w") as f:
        f.write("8000000\n")
        for i in range(1, 8000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
#Finally..
%%time
!python max_pairwise_product_sort.py < input16.txt

63999992000000
CPU times: user 36.2 ms, sys: 976 µs, total: 37.1 ms
Wall time: 2.62 s


###Correctness Tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_sort.cpp -o max_pairwise_product_long

for i in {0..10}
do
    ./max_pairwise_product_sort < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_sort.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_ef.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    long long p_max = 0;                     //1
    long long s_max = 0;                     //1
    for(int i = 0; i < n; ++i) {             //1 + 2n
        if (v[i] > s_max){                    //n
            s_max = v[i];                    //[0, n]
        }
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            p_max = v[i];                    //[0, n]
        }
    }
    return (long long) p_max * s_max;       //2 -> 9 + 4n + 3[0, n]           //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    std::cin >> n;                          //1
    std::vector<int> v(n);                  //n
    for(int i = 0; i < n; i++){             //1 + 2n
        std::cin >> v[i];                   //n
    }
    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n
    return 0;
}

Overwriting max_pairwise_product_ef.cpp


In [ ]:
%%writefile max_pairwise_product_ef.py
#Python3
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product_f(l))

Overwriting max_pairwise_product_ef.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!./max_pairwise_product_ef < input1.txt
!./max_pairwise_product_ef < input2.txt

90
90


In [ ]:
!python max_pairwise_product_ef.py < input1.txt
!python max_pairwise_product_ef.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input3.txt

40000000000

real	0m0.003s
user	0m0.003s
sys	0m0.000s


In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input11.txt

39999800000

real	0m0.072s
user	0m0.067s
sys	0m0.003s


In [ ]:
%%time
!python max_pairwise_product_ef.py < input3.txt

40000000000
CPU times: user 7.03 ms, sys: 0 ns, total: 7.03 ms
Wall time: 107 ms


In [ ]:
%%time
!python max_pairwise_product_ef.py < input11.txt

39999800000
CPU times: user 9.99 ms, sys: 982 µs, total: 11 ms
Wall time: 306 ms


20 stress tests

In [ ]:
%%writefile stress_tests_ef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();
    ins++;
    long long p_max = 0;
    ins++;
    long long s_max = 0;
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {
        ins += 2;
        ins++;
        if (v[i] > s_max){
            s_max = v[i];
            ins++;
        }
        ins++;
        if (v[i] > p_max){
            s_max = p_max;
            ins++;
            p_max = v[i];
            ins++;
        }
    }

    return (long long) p_max * s_max;
}

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 69 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::cout << "n: " << n << " Product: " << max_pairwise_product_f(v) << '\n';
    }
    return 0;
}

Overwriting stress_tests_ef_cpp.cpp


In [ ]:
%%writefile stress_tests_ef_py.py
#Python3
import random
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        l = [random.randint(17, 24531) for i in range(n)]
        print(f"n: {n} Product: {max_pairwise_product_f(l)}")

Overwriting stress_tests_ef_py.py


In [ ]:
!g++ stress_tests_ef_cpp.cpp -o stress_tests_ef_cpp
!./stress_tests_ef_cpp

n: 36 Product: 863968
n: 57 Product: 980036
n: 60 Product: 898343
n: 31 Product: 927682
n: 13 Product: 762600
n: 51 Product: 954088
n: 12 Product: 683388
n: 26 Product: 981034
n: 70 Product: 960364
n: 14 Product: 858552
n: 61 Product: 993010
n: 52 Product: 963270
n: 10 Product: 631764
n: 32 Product: 994008
n: 21 Product: 960319
n: 24 Product: 828625
n: 2 Product: 74562
n: 54 Product: 904397
n: 8 Product: 886974
n: 23 Product: 911064


In [ ]:
!python stress_tests_ef_py.py

n: 46 Product: 592750422
n: 24 Product: 457568320
n: 79 Product: 571651534
n: 19 Product: 486421504
n: 43 Product: 493548572
n: 88 Product: 596595348
n: 2 Product: 233067625
n: 72 Product: 597068784
n: 49 Product: 549979632
n: 15 Product: 532085705
n: 89 Product: 594091392
n: 59 Product: 596725888
n: 67 Product: 580249645
n: 53 Product: 559629572
n: 64 Product: 562290780
n: 17 Product: 517944606
n: 38 Product: 533143080
n: 18 Product: 562673741
n: 30 Product: 579036030
n: 27 Product: 581123252


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_ef_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    ins++;
    long long p_max = 0;                     //1
    ins++;
    long long s_max = 0;                     //1
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {             //1 + 2n
        ins += 2;
        ins++;
        if (v[i] > s_max){                    //n
            s_max = v[i];                     //[0, n]
            ins++;
        }
        ins++;
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            ins++;
            p_max = v[i];                    //[0, n]
            ins++;
        }
    }
    ins += 2;
    return (long long) p_max * s_max;       //2 -> 6 + 4n + 3[0, n]          //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    ins++;
    std::cin >> n;                          //1
    ins++;
    std::vector<int> v(n);                  //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){             //1 + 2n
        ins += 2;
        std::cin >> v[i];                   //n
        ins++;
    }

    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n + 6 + 4n + 3[0, n] -> 10 +8n + 3[0, n] -> 10 + 11n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Overwriting max_pairwise_product_ef_ins.cpp


In [ ]:
%%writefile max_pairwise_product_ef_ins_py.py
#Python3
ins = 0
def max_pairwise_product_f(l):
    global ins
    p_max = 0                               #1
    ins += 1
    s_max = 0                               #1
    ins += 1
    ins += 1
    for i in l:                             #1 + 2n
        ins += 2
        if i > s_max:                       #n
            s_max = i                       #[0, n]           #[0, n] means that, in the worst case, executes n times
            ins += 1
        if i > p_max:                       #n
            s_max = p_max                   #[0, n]
            ins += 1
            p_max = i                       #[0, n]
            ins += 1

    ins += 2
    return p_max * s_max                    #2       #5 + 4n + 3[0, n] -> 5 + 7n worst

if __name__ == "__main__":
        n = int(input())                                 #1
        ins += 1
        l = list(map(int, input().split()))              #5n
        ins += 5 * n
        ins += 1
        print(max_pairwise_product_f(l))                 #1
        print(f"Executed instructions: {ins}")           #-> 2 + 5n + 5 + 7n -> 7 + 12n

Overwriting max_pairwise_product_ef_ins_py.py


In [ ]:
!g++ max_pairwise_product_ef_ins.cpp -o max_pairwise_product_ef_ins
!./max_pairwise_product_ef_ins < input1.txt
!./max_pairwise_product_ef_ins < input2.txt
!./max_pairwise_product_ef_ins < input3.txt

90
Executed instructions: 119
90
Executed instructions: 93
40000000000
Executed instructions: 68


In [ ]:
!time ./max_pairwise_product_ef_ins < input11.txt

39999800000
Executed instructions: 2200009

real	0m0.148s
user	0m0.115s
sys	0m0.004s


In [ ]:
!python max_pairwise_product_ef_ins_py.py < input1.txt
!python max_pairwise_product_ef_ins_py.py < input2.txt

90
Executed instructions: 107
90
Executed instructions: 81


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input3.txt

40000000000
Executed instructions: 60
CPU times: user 6.98 ms, sys: 0 ns, total: 6.98 ms
Wall time: 108 ms


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input11.txt

39999800000
Executed instructions: 2000007
CPU times: user 6.38 ms, sys: 1.06 ms, total: 7.44 ms
Wall time: 205 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.006/ 2200009
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

2.727261570293576e-09
366668166.6666667


###Correctness Tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

for i in {0..10}
do
    ./max_pairwise_product_ef_cpp < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_ef.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

###Estimated maximum that can be solved in 1 second
$$(10 + 11n)*1.47*10^{-9} = 1s$$
$$ n \approx 60'000.000 $$

In [ ]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 60000000 <<"\n";  // n
 for (int i=1; i <60000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

In [ ]:
!time ./max_pairwise_product_ef_cpp < input15.txt #The arrays are being saved in the stack. This gives the "practical" time instead of the theoretical one.

In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("60000000\n")
        for i in range(1, 60000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
%%time
!python max_pairwise_product_ef.py < input14.txt #Readings -> slow code

3599999940000000
CPU times: user 397 ms, sys: 58.7 ms, total: 455 ms
Wall time: 37.1 s


### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [ ]:
%%writefile max_pairwise_product_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    int n;
    std::cin >> n;
    int p_max = INT_MIN;
    int s_max = INT_MIN;

    for (int i = 0; i < n; i++){
        int x;
        std::cin >> x;
        if(x > p_max){
            s_max = p_max;
            p_max = x;
        }
        else if(x > s_max){
            s_max = x;
        }
    }

    std::cout << (long long) p_max * s_max << "\n";
    return 0;
}

Writing max_pairwise_product_gef_cpp.cpp


Test with inputs 1 and 2




In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!./max_pairwise_product_gef_cpp < input1.txt
!./max_pairwise_product_gef_cpp < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input3.txt

40000000000

real	0m0.004s
user	0m0.001s
sys	0m0.002s


In [ ]:
!time ./max_pairwise_product_gef_cpp < input11.txt

39999800000

real	0m0.122s
user	0m0.093s
sys	0m0.006s


20 stress tests

In [ ]:
%%writefile stress_tests_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 77 + 2;
        int p_max = INT_MIN;
        int s_max = INT_MIN;

        for (int i = 0; i < n; i++){
            int x = rand() % 1000 + 2;
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }

        std::cout << "n: " << n << " Product: " << (long long) p_max * s_max << "\n";
    }
    return 0;
}

Overwriting stress_tests_gef_cpp.cpp


In [ ]:
!g++ stress_tests_gef_cpp.cpp -o stress_tests_gef_cpp
!./stress_tests_gef_cpp

n: 52 Product: 863968
n: 59 Product: 956084
n: 53 Product: 898343
n: 22 Product: 927682
n: 2 Product: 56643
n: 32 Product: 928140
n: 48 Product: 885256
n: 26 Product: 981034
n: 40 Product: 928812
n: 23 Product: 876600
n: 55 Product: 991020
n: 27 Product: 990016
n: 11 Product: 557996
n: 70 Product: 994008
n: 53 Product: 961308
n: 58 Product: 926316
n: 39 Product: 979080
n: 3 Product: 290360
n: 23 Product: 876015
n: 26 Product: 874290


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                          //1
    ins++;
    std::cin >> n;                  //1
    ins++;
    int p_max = INT_MIN;            //1
    ins++;
    int s_max = INT_MIN;            //1
    ins++;
    ins++;
    for (int i = 0; i < n; i++){    //1 + 2n
        ins += 2;
        int x;                      //n
        ins++;
        std::cin >> x;              //n
        ins++;
        ins += 2; //both ifs
        if(x > p_max){              //[0, n]           //[0, n] means that, in the worst case, executes n times.
            s_max = p_max;          //[0, n]
            ins++;
            p_max = x;              //[0, n]
            ins++;
        }
        else if(x > s_max){         //[0, n]
            s_max = x;              //[0, n]
            ins++;
        }
    }

    ins += 3;
    std::cout << (long long) p_max * s_max << "\n"; //3 -> 8 + 4n + 5[0, n] -> 8 + 9n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_gef_ins_cpp.cpp


In [ ]:
!g++ max_pairwise_product_gef_ins_cpp.cpp -o max_pairwise_product_gef_ins_cpp
!./max_pairwise_product_gef_ins_cpp < input1.txt
!./max_pairwise_product_gef_ins_cpp < input2.txt
!./max_pairwise_product_gef_ins_cpp < input3.txt

90
Executed instructions: 88
90
Executed instructions: 71
40000000000
Executed instructions: 52


In [ ]:
!time ./max_pairwise_product_gef_ins_cpp < input11.txt

39999800000
Executed instructions: 1600008

real	0m0.084s
user	0m0.077s
sys	0m0.003s


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.0071 / 1600008
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.437477812610937e-09
225353239.4366197


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_gef_py.py
#Python3
if __name__ == "__main__":
    n = int(input())
    p_max = -1
    s_max = -1
    for i in input().split():
        i = int(i)
        if(i > p_max):
            s_max = p_max
            p_max = i
        elif i > s_max:
            s_max = i
    print(p_max * s_max)

Writing max_pairwise_product_gef_py.py


Test with inputs 1 and 2




In [ ]:
!python max_pairwise_product_gef_py.py < input1.txt
!python max_pairwise_product_gef_py.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input3.txt

40000000000
CPU times: user 6.92 ms, sys: 0 ns, total: 6.92 ms
Wall time: 107 ms


In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input11.txt

39999800000
CPU times: user 7.52 ms, sys: 20 µs, total: 7.54 ms
Wall time: 206 ms


20 stress tests

In [ ]:
%%writefile stress_tests_gef_py.py
import random
#Python3
if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        p_max = -1
        s_max = -1
        for x in range(n):
            i = random.randint(19, 133700)
            if(i > p_max):
                s_max = p_max
                p_max = i
            elif i > s_max:
                s_max = i
        print(f"n: {n} Product: {p_max * s_max}")

Writing stress_tests_gef_py.py


In [ ]:
!python stress_tests_gef_py.py

n: 73 Product: 16702853320
n: 66 Product: 16020152225
n: 62 Product: 16807892775
n: 37 Product: 17519059044
n: 27 Product: 14415461899
n: 45 Product: 15561990672
n: 85 Product: 17234607936
n: 90 Product: 17453135590
n: 95 Product: 16847850866
n: 46 Product: 16857701025
n: 85 Product: 16878327752
n: 68 Product: 16643544120
n: 18 Product: 17165625200
n: 50 Product: 15311336250
n: 84 Product: 16195114080
n: 15 Product: 15789575550
n: 79 Product: 16002465561
n: 63 Product: 17679012126
n: 40 Product: 17325850245
n: 45 Product: 16590511800


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_py.py
import random
#Python3
if __name__ == "__main__":
        ins = 0
        n = int(input())                 #1
        ins += 1
        p_max = -1                       #1
        ins += 1
        s_max = -1                       #1
        ins += 1
        ins += 1
        for i in input().split():        #1 + 3n (n reads, 2n comparisons unitl end of input)
            ins += 2
            i = int(i)                   #n
            ins += 1
            ins += 2
            if(i > p_max):               #[0, n]
                s_max = p_max            #[0, n]
                ins += 1
                p_max = i                #[0, n]
                ins += 1
            elif i > s_max:              #[0, n]
                s_max = i                #[0, n]
                ins += 1
        print(p_max * s_max)             #2
        print(f"Executed instructions: {ins}") #->6 + 4n + 5[0, n] -> 6 + 9n worst case

Writing max_pairwise_product_gef_ins_py.py


In [ ]:
!python max_pairwise_product_gef_ins_py.py < input1.txt
!python max_pairwise_product_gef_ins_py.py < input2.txt
!python max_pairwise_product_gef_ins_py.py < input3.txt

90
Executed instructions: 74
90
Executed instructions: 57
40000000000
Executed instructions: 42


In [ ]:
%%time
!python max_pairwise_product_gef_ins_py.py < input11.txt

39999800000
Executed instructions: 1400004
CPU times: user 7.33 ms, sys: 1.06 ms, total: 8.39 ms
Wall time: 308 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.0086 / 1400004
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

6.14283959188688e-09
162791162.79069766


###Correctness tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp

for i in {0..10}
do
    ./max_pairwise_product_gef_cpp < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_gef_py.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

###Estimated maximum that can be solved in 1 second
$$(6 + 9n)*1.47*10^{-9} = 1s$$
$$ n \approx 75'000.000 $$

In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 75000000 <<"\n";  // n
 for (int i=1; i <75000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest

In [ ]:
!./WriteTest

In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input15.txt #The theoretical max in 1 sec is 75M. The readings in disk pf the .txt files degradate this time.

5624999925000000

real	0m31.962s
user	0m31.397s
sys	0m0.281s


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("75000000\n")
        for i in range(1, 75000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input14.txt #Again, the theoretical holds. The readings from disk degradate performance.

5624999925000000
CPU times: user 375 ms, sys: 51.1 ms, total: 426 ms
Wall time: 42.8 s
